requests-for-comments/2021-10-05-group-analytics.md

```diff
@@ -0,0 +1,218 @@
+
+## Overview
+
+This RFC proposes a solution which would allow PostHog users to leverage insights and feature flags not only by users but also by groups (e.g. organizations, projects).
+
+## Background
+B2B products don’t deal directly with users, but with other companies and hence have more complicated needs around what kinds of metrics are most important for them.
+
+
+Lack of this feature has been a major driver for users wanting data exports in different analytics products, forcing users to roll their own analytics on top of data warehouses to get at their key metrics.
+
+### Example usecases
+
+
+- Show me how many active “organizations” there are
+- Show only pageviews from organizations who are paying us (filtering by group properties)
+- Have an overview of groups and their relations with users
+- Feature flags: roll out feature group-by-group
+
+For PostHog the product, the 3 group types we would use are:
+
+- Instances (e.g. cloud, posthog.someproduct.org, …)
+- Organizations
+- Projects
+
+## Semantics & Constraints
+
+### 1. Users can belong to multiple groups
+
+Some examples:
+
+1. Contractor type work (e.g. for farm management agronomists work for multiple farms at once)
+2. Billing happens on a different level (e.g. our projects vs organization)
+3. Users clients are global companies with a lot of local “branches” where execs need access to everything
+4. Internal users have access to debug issues on multiple accounts
+
+### 2. User is active in only “group” at any point in time
+
+Whenever an event is made, we can (and should) make the assumption that it belongs to a single group (e.g. organization). This is needed to avoid “double-counting” in stats.
+
+Example: If Yakko has been added to 20 different organizations and is active, we don’t want to count all of them as active.
+
+However users can be in a different "group type" at the same time.
+
+Example: Yakkos events can be counted to be both in Organization posthog as well as the cloud instance at the same time.
+
+### 3. Limited amount of group types
+
+From speaking with clients, their internal amount of different “group types” (e.g. `organization`) is limited, though the number of groups under a group type is unlimited (e.g. Facebook if it was an organization).
+
+This constraint allows us to implement things slightly more efficiently and have a nicer UX compared to unbounded amount of groups.
+
+### 4. Feature flags in groups should be consistent
+
+For the users who are using feature flags, they want to ensure a consistent experience within a group during rollout.
+
+This is to avoid situations where some users see a new feature and some do not.
+
+This also factors into their later analysis - for example, they’d like to treat two different “organizations” both as a single datapoint even if one has 1500 users and the other has 1.
+
+## Proposal
+
+### Groups are event-level
```

Comment:
One thing is not clear to me from a UX perspective is how the user will understand the group hierarchy conceptually. For us an instance can have multiple orgs, an org multiple teams, ... but our users may have a different setup with a different hierarchy. I want to make sure we have a lot of clarity around the mental model of this from a user's perspective.

Reply:
We asked about this in the interviews: the TL;DR is that it didn't seem to be an urgent issue. Users often did have a hierarchical structure to their groups which they knew from heart and didn't need help on. 

But as with everything, we might need to iterate this after shipping.